src/Common/src/System/Net/Sockets/SocketAPMExtensions.cs

```diff
@@ -829,6 +829,21 @@ public static int EndSendTo(this Socket socket, IAsyncResult asyncResult)
         {
             return TaskToApm.End<int>(asyncResult);
         }
+
+        // Behavior adapter.
+        private static ArraySegment<byte> CreateArraySegment(byte[] buffer, int offset, int size)
```

Comment:
@stephentoub @josguil  Taking a second look, to correctly match the APM behavior, the code should actually be.
Please note that we're not trying to have the best performance for the adapter but existing code should work with small or little modification (except for cases where it was hard or impossible to achieve the old behavior, as outlined in the test changes).

```C#
            // Validate input parameters.
            if (buffer == null)
            {
                throw new ArgumentNullException("buffer");
            }
            if (offset < 0 || offset > buffer.Length)
            {
                throw new ArgumentOutOfRangeException("offset");
            }
            if (size < 0 || size > buffer.Length - offset)
            {
                throw new ArgumentOutOfRangeException("size");
            }

           return new ArraySegment<byte>(buffer, offset, size);
```